Code/GraphMol/FMCS/FMCS.cpp

```diff
@@ -236,15 +245,54 @@ bool checkAtomChirality(const MCSAtomCompareParameters& p,
   return true;
 }
 
+bool checkAtomDistance(const MCSAtomCompareParameters& p,
+                       const ROMol& mol1, unsigned int atom1,
+                       const ROMol& mol2, unsigned int atom2,
+                       MCSCompareFunctionsData& cfd){
+    unsigned int confId1 = cfd.conformerIdxMap[&mol1];
+    unsigned int confId2 = cfd.conformerIdxMap[&mol2];
+    // If there is no setting for conformer idxs, default to 0
+    if (mol1.getNumConformers() <= confId1 ||
+        mol2.getNumConformers() <= confId2){
+      throw std::runtime_error(
+        "FMCS. Invalid argument. requested conformer unavailable");
+    }
+    const Atom* atm1 = mol1.getAtomWithIdx(atom1);
+    const Atom* atm2 = mol2.getAtomWithIdx(atom2);
+    const Atom* mapAtmFirst = atm1<atm2 ? atm1 : atm2;
+    const Atom* mapAtmSecond = atm1<atm2 ? atm2 : atm1;
+    MCSAtomDistanceCache::iterator cacheIt;
+    std::map<const Atom*, bool>::iterator innerMapIt;
+
+    cacheIt = cfd.atomDistanceCache.find(mapAtmFirst);
+    if (cacheIt != cfd.atomDistanceCache.end()){
+      innerMapIt = cacheIt->second.find(mapAtmSecond);
+      if (innerMapIt != cacheIt->second.end()){
+        return innerMapIt->second;
+      }
+    }
+
+    const Conformer &ci1 = mol1.getConformer(confId1);
+    const Conformer &ci2 = mol2.getConformer(confId2);
+    const RDGeom::Point3D &pos1 = ci1.getAtomPos(atom1);
+    const RDGeom::Point3D &pos2 = ci2.getAtomPos(atom2);
+    bool withinRange = (pos1 - pos2).length() <= p.MaxDistance;
+    cfd.atomDistanceCache[mapAtmFirst][mapAtmSecond] = withinRange;
+    return withinRange;
+}
+
 bool MCSAtomCompareAny(const MCSAtomCompareParameters& p, const ROMol& mol1,
                        unsigned int atom1, const ROMol& mol2,
-                       unsigned int atom2, void*) {
+                       unsigned int atom2, MCSCompareFunctionsData& cfd) {
```

Comment:
This is a breaking change - if anyone has external C++ code using custom compare functors, they will need to  change the signature and then use the cfd.userData member instead. I'll justify this a little later.

Reply:
This is the obvious way to address `userData` cleanly. However it will break all existing code relying on custom MCS functors. This is why I went for the complicated route that you can see in the current code, which although not elegant, preserves backwards compatibility and still allows to use `userData` for caching purposes.
Let’s see what others think. When I implemented the custom Python functors I would have loved to do that exact code change, but I refrained from doing it as it is a big breaking change.

Reply:
Thanks -
I'm putting together a version with less aggressive changes so I'm not creating so much review work for other people.